### PR TITLE
task3

### PR DESCRIPTION
**Pull request by ZbigniewBockowski**

### CODE REVIEW

File: src/fight_club_api.py [hunk 1]
```diff
@@ -33,6 +39,40 @@ def __fight(self, fighters, max_rounds_number):
...
+            resp.body = json.dumps({'error': [err.args]}) 
+            return
+
 api = falcon.API()
+resource = FightClubApi()
 
-api.add_route('/fight', FightClubApi())
\ No newline at end of file
+api.add_route('/fight', resource)
+api.add_route('/fight', resource)
```
Comment (fancysnake) [inline src/fight_club_api.py]: Why double definition here?

File: src/fight_club_api.py [hunk 2]
```diff
@@ -33,6 +39,40 @@ def __fight(self, fighters, max_rounds_number):
...
+        try:
+            fights = FightClubDbService.find_fights_by_params(req.params.items())
+            resp.content_type = 'application/json'
+            resp.body = jsonpickle.encode(fights, unpicklable=False)
+        except AttributeError as err:
+            resp.status = falcon.HTTP_400
+            resp.body = json.dumps({'error': [err.args]}) 
+            return
+
+    def on_get_fighter(self, req, resp):
```
Comment (fancysnake) [inline src/fight_club_api.py]: Please move this and round to separate REST resources.

File: src/fight_club_db_service.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	@staticmethod
+	def find_fighter_by_index(index: str) -> 'Fighter':
+		fighter = Fighter.get_or_none(Fighter.index == index)
+		return fighter
+
+	@staticmethod
+	def save_fighter(fighter_index: str, dto: 'FighterDto'):
+		Fighter.create(index=fighter_index, name=dto.name, hp=dto.hp, damage_dice=dto.damage_dice, damage_bonus=dto.damage_bonus)
```
Comment (fancysnake) [inline src/fight_club_db_service.py]: For short:
```python
Fighter.create(index=fighter_index, **dataclasses.asdict(dto))
```

File: src/fight_club_db_service.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
+from model import *
+import operator
+from functools import reduce
+
+class FightClubDbService:
+	fighter_parameters_map = {
```
Comment (fancysnake) [inline src/fight_club_db_service.py:6]: Check `Fighter._meta.get_field_names()`

File: src/fight_club_db_service.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		fight = Fight.create(winner=fight_result.victory.winner_name, rounds_number=fight_result.victory.rounds_number, ko=fight_result.victory.ko)
+		for round in fight_result.rounds:
+			Round.create(fight_id=fight, round_nr=round.round_number, attacker=round.attacker, defender=round.defender, damage=round.damage, previous_hp=round.previous_hp, current_hp=round.current_hp)
+
+	@staticmethod
+	def find_fighters_by_params(params):
+		predicates = []
+		for key, value in params:
+			if key not in FightClubDbService.fighter_parameters_map:
+				raise AttributeError(f"Parameter {key} is unknown for Fighter object."
```
Comment (fancysnake) [inline src/fight_club_db_service.py:50]: Please rise a custom exception here

File: src/fight_club_db_service.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+			if key not in FightClubDbService.fight_parameters_map:
+				raise AttributeError(f"Parameter {key} is unknown for Fight object."
+					f" Following attributes are allowed: {', '.join(FightClubDbService.fight_parameters_map)}")
+			if key is not None and value is not None:
+				predicates.append((FightClubDbService.fight_parameters_map[key]==value))
+		query = (Round.select())
+		if len(predicates):
+			query = (query.where(reduce(operator.and_, predicates)))
+		fights = []
+		for fight in query:
```
Comment (fancysnake) [inline src/fight_club_db_service.py]: `fighters=list(query)` should be good here

**Review by fancysnake — CHANGES_REQUESTED** [review]

File: src/fight_club_db_service.py (hunk 4 above)
Comment (ZbigniewBockowski) [reply src/fight_club_db_service.py:6]: I need references to static model properties. Unfortunately field names are not helpful there.

File: src/fight_club_db_service.py (hunk 5 above)
Comment (ZbigniewBockowski) [reply src/fight_club_db_service.py:50]: Why different if that one is intended to handle such cases?

File: src/fight_club_db_service.py (hunk 4 above)
Comment (fancysnake) [reply src/fight_club_db_service.py:6]: try `getattr(Fighter, field_name)` then

File: src/fight_club_db_service.py (hunk 5 above)
Comment (fancysnake) [reply src/fight_club_db_service.py:50]: ok

File: src/fight_club_db_service.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		predicates = []
+		for key, value in params:
+			if key not in FightClubDbService.fighter_parameters_map:
+				raise AttributeError(f"Parameter {key} is unknown for Fighter object."
+					f" Following attributes are allowed: {', '.join(FightClubDbService.fighter_parameters_map)}")
+			if key is not None and value is not None:
+				predicates.append((FightClubDbService.fighter_parameters_map[key]==value))
+		query = (Round.select())
+		if len(predicates):
+			query = (query.where(reduce(operator.and_, predicates)))
```
Comment (fancysnake) [inline src/fight_club_db_service.py:56]: :heart:

File: src/fight_club_db_service.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	def save_fight_result(fight_result: 'FightResultDto'):
+		fight = Fight.create(winner=fight_result.victory.winner_name, rounds_number=fight_result.victory.rounds_number, ko=fight_result.victory.ko)
+		for round in fight_result.rounds:
+			Round.create(fight_id=fight, round_nr=round.round_number, attacker=round.attacker, defender=round.defender, damage=round.damage, previous_hp=round.previous_hp, current_hp=round.current_hp)
+
+	@staticmethod
+	def find_fighters_by_params(params):
+		predicates = []
+		for key, value in params:
+			if key not in FightClubDbService.fighter_parameters_map:
```
Comment (fancysnake) [inline src/fight_club_db_service.py:49]: `if not hasattr(Fighter, key)`

File: src/fight_club_db_service.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+				raise AttributeError(f"Parameter {key} is unknown for Fighter object."
+					f" Following attributes are allowed: {', '.join(FightClubDbService.fighter_parameters_map)}")
+			if key is not None and value is not None:
+				predicates.append((FightClubDbService.fighter_parameters_map[key]==value))
```
Comment (fancysnake) [inline src/fight_club_db_service.py:53]: `predicates.append(getattr(Fighter, key) == value)`

**Review by fancysnake — CHANGES_REQUESTED** [review]

**Pull request merged**